Mathlib/MeasureTheory/Order/UpperLower.lean

```diff
@@ -0,0 +1,206 @@
+
+#align_import measure_theory.order.upper_lower from "leanprover-community/mathlib"@"b1abe23ae96fef89ad30d9f4362c307f72a55010"
+
+/-!
+# Order-connected sets are null-measurable
+
+This file proves that order-connected sets in `ℝⁿ` under the pointwise order are null-measurable.
+Recall that `x ≤ y` iff `∀ i, x i ≤ y i`, and `s` is order-connected iff
+`∀ x y ∈ s, ∀ z, x ≤ z → z ≤ y → z ∈ s`.
+
+## Main declarations
+
+* `Set.OrdConnected.null_frontier`: The frontier of an order-connected set in `ℝⁿ` has measure `0`.
+
+## Notes
+
+We prove null-measurability in `ℝⁿ` with the `∞`-metric, but this transfers directly to `ℝⁿ` with
+the Euclidean metric because they have the same measurable sets.
+
+Null-measurability can't be strengthened to measurability because any antichain (and in particular
+any subset of the antidiagonal `{(x, y) | x + y = 0}`) is order-connected.
+
+## TODO
+
+Generalize so that it also applies to `ℝ × ℝ`, for example.
+-/
+
+open Filter MeasureTheory Metric Set
+open scoped Topology
+
+namespace Set
+variable {ι : Type*} {α : ι → Type*} {i : ι} {s : Set ι} {t : ∀ i, Set (α i)}
+
+-- lemma eval_image_pi_of_not_mem (hi : i ∉ s) : Function.eval i '' s.pi t = univ :=
+--   eq_univ_of_forall (by simp)
+
+end Set
+
+section
+variable {α : Type*} [TopologicalSpace α] [Preorder α] {x : α}
+
+instance OrderDual.instNeBotNhdsWithinIoi [(𝓝[<] x).NeBot] : (𝓝[>] toDual x).NeBot := ‹_›
+instance OrderDual.instNeBotNhdsWithinIio [(𝓝[>] x).NeBot] : (𝓝[<] toDual x).NeBot := ‹_›
+
+end
+
+section
+variable {ι : Type*} {π : ι → Type*} [∀ i, TopologicalSpace (π i)] [∀ i, Preorder (π i)]
+open TopologicalSpace
+
+-- lemma isOpenMap_eval (i : ι) : IsOpenMap (Function.eval i : (∀ i, π i) → π i) := by
+--   refine (isTopologicalBasis_pi fun _ ↦ isTopologicalBasis_opens).isOpenMap_iff.2 ?_
+--   rintro _ ⟨U, s, hU, rfl⟩
+--   obtain h | h := ((s : Set ι).pi U).eq_empty_or_nonempty
+--   · simp [h]
+--   by_cases hi : i ∈ s
+--   · rw [eval_image_pi (mod_cast hi) h]
+--     exact hU _ hi
+--   · sorry
+
+open OrderDual
+
+instance Pi.instNeBotNhdsWithinIio [Nonempty ι] (x : ∀ i, π i) [hx : ∀ i, (𝓝[<] x i).NeBot] :
+    (𝓝[<] x).NeBot := by
+  -- Let's show that every neighborhood of `x` intersect `Iio x`.
+  simp_rw [nhdsWithin_neBot,
+  -- WLOG the neighborhood is the product of neighborhoods `V i` of `x i`.
+    (isTopologicalBasis_pi fun _ ↦ isTopologicalBasis_opens).mem_nhds_iff] at hx ⊢
+  rintro t ⟨U, ⟨U, s, hU, rfl⟩, hxU, hUt⟩
+  classical
+  -- `isTopologicalBasis_pi` does not give us a product of open sets but a product of open sets on a
+  -- finset, so we need the following nonsense to make sure every factor is open.
+  let V (i) := if i ∈ s then U i else univ
+  have hV (i) : IsOpen (V i) := by
+    unfold_let; dsimp; split_ifs with hi; exacts [hU _ hi, isOpen_univ]
+  have hVU : pi s V = pi s U := Set.pi_congr rfl fun i hi ↦ if_pos hi
+  have hxV (i) : x i ∈ V i := by unfold_let; dsimp; split_ifs with hi; exacts [hxU _ hi, mem_univ _]
```

Comment:
To the topologists (@PatrickMassot, @ADedecker ?), why is `isTopologicalBasis_pi` giving me those open sets whose projections are not all open? This is mathematically correct but really impractical as you see here.

Reply:
Moved to #13642